File: src/main/java/com/hazza/LeetCodeByJava/No65ValidNumber.java
package com.hazza.LeetCodeByJava;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/valid-number/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-20
 * Time: 9:23 AM
 */
public class No65ValidNumber {
    public boolean isNumber(String s) {
        s = s.trim();
        int len = s.length();
        boolean numberFlag = false;
        boolean eFlag = false;
        boolean pointFlag = false;
        boolean numberAfterEFlag = false;

        for (int i = 0; i < len; i++) {
            if (s.charAt(i) >= '0' && s.charAt(i) <= '9') {
                numberFlag = true;
                numberAfterEFlag = true;
            } else if (s.charAt(i) == 'e') {
                if (!numberFlag || eFlag) return false;
                eFlag = true;
                numberAfterEFlag = false;
            } else if (s.charAt(i) == '.') {
                if (eFlag || pointFlag) return false;
                pointFlag = true;
            } else if (s.charAt(i) == '+' || s.charAt(i) == '-') {
                if (i != 0 && s.charAt(i - 1) != 'e') return false;
            } else return false;
        }

        return numberFlag && numberAfterEFlag;
    }
}

File: src/main/python/leetcode_by_python/SegmentTwo/No65_Valid_Number.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/15 20:07
# https://leetcode.com/problems/valid-number/


class Solution:
    def isNumber(self, s: str) -> bool:
        s = s.strip()
        numberFlag = False
        eFlag = False
        pointFlag = False
        numberAfterEFlag = False

        for i in range(len(s)):
            if '0' <= s[i] <= '9':
                numberFlag = True
                numberAfterEFlag = True
            elif s[i] == 'e':
                if not numberFlag or eFlag:
                    return False
                eFlag = True
                numberAfterEFlag = False
            elif s[i] == '.':
                if pointFlag or eFlag:
                    return False
                pointFlag = True
            elif s[i] == '-' or s[i] == '+':
                if i != 0 and s[i - 1] != 'e':
                    return False
            else:
                return False

        return numberFlag and numberAfterEFlag


if __name__ == '__main__':
    solution = Solution()

    assert solution.isNumber('0') is True
    assert solution.isNumber(' 0.1 ') is True
    assert solution.isNumber('abc') is False
    assert solution.isNumber('1 a') is False
    assert solution.isNumber('2e10') is True
    assert solution.isNumber(' -90e3   ') is True
    assert solution.isNumber(' 1e') is False
    assert solution.isNumber('e3') is False
    assert solution.isNumber(' 6e-1') is True
    assert solution.isNumber(' 99e2.5 ') is False
    assert solution.isNumber(' --6 ') is False
    assert solution.isNumber('53.5e93') is True
    assert solution.isNumber('-+3') is False
    assert solution.isNumber('95a54e53') is False